server/cluster/hot_region_storage.go

```diff
@@ -0,0 +1,348 @@
+	"github.com/pingcap/log"
+	"github.com/syndtr/goleveldb/leveldb"
+	"github.com/syndtr/goleveldb/leveldb/iterator"
+	"github.com/syndtr/goleveldb/leveldb/util"
+	"github.com/tikv/pd/pkg/encryption"
+	"github.com/tikv/pd/pkg/errs"
+	"github.com/tikv/pd/server/encryptionkm"
+	"github.com/tikv/pd/server/kv"
+	"github.com/tikv/pd/server/member"
+	"github.com/tikv/pd/server/statistics"
+)
+
+// HotRegionStorage is used to storage hot region info,
+// It will pull the hot region information according to the pullInterval interval.
+// And delete and save data beyond the remainingDays.
+// Close must be called after use.
+type HotRegionStorage struct {
+	*kv.LeveldbKV
+	encryptionKeyManager *encryptionkm.KeyManager
+	mu                   sync.RWMutex
+	batchHotInfo         map[string]*statistics.HistoryHotRegion
+	remianedDays         int64
+	pullInterval         time.Duration
+	compactionCountdown  int
+	hotRegionInfoCtx     context.Context
+	hotRegionInfoCancel  context.CancelFunc
+	cluster              *RaftCluster
+	member               *member.Member
+}
+
+const (
+	// leveldb will run compaction after 30 times delete.
+	defaultCompactionTime = 30
+	// delete will run at this o`clock.
+	defaultDeleteTime = 4
+)
+
+// HotRegionTypes stands for hot type.
+var HotRegionTypes = []string{
+	"read",
+	"write",
+}
+
+// NewHotRegionsStorage create storage to store hot regions info.
+func NewHotRegionsStorage(
+	ctx context.Context,
+	path string,
+	encryptionKeyManager *encryptionkm.KeyManager,
+	cluster *RaftCluster,
+	member *member.Member,
+	remianedDays int64,
+	pullInterval time.Duration,
+) (*HotRegionStorage, error) {
+	levelDB, err := kv.NewLeveldbKV(path)
+	if err != nil {
+		return nil, err
+	}
+	hotRegionInfoCtx, hotRegionInfoCancle := context.WithCancel(ctx)
+	h := HotRegionStorage{
+		LeveldbKV:            levelDB,
+		encryptionKeyManager: encryptionKeyManager,
+		batchHotInfo:         make(map[string]*statistics.HistoryHotRegion),
+		remianedDays:         remianedDays,
+		pullInterval:         pullInterval,
+		compactionCountdown:  defaultCompactionTime,
+		hotRegionInfoCtx:     hotRegionInfoCtx,
+		hotRegionInfoCancel:  hotRegionInfoCancle,
+		cluster:              cluster,
+		member:               member,
+	}
+	if remianedDays > 0 {
+		h.backgroundFlush()
+		h.backgroundDelete()
+	}
+	return &h, nil
+}
+
+// delete hot region which update_time is smaller than time.Now() minus remain day in the background.
+func (h *HotRegionStorage) backgroundDelete() {
+	// make delete happened in defaultDeleteTime clock.
+	now := time.Now()
+	next := now.Add(time.Hour * 24)
+	next = time.Date(next.Year(), next.Month(), next.Day(), defaultDeleteTime, 0, 0, 0, next.Location())
+	t := time.NewTicker(next.Sub(now))
+	go func() {
+		select {
+		case <-t.C:
+
+		case <-h.hotRegionInfoCtx.Done():
+			return
+		}
+		ticker := time.NewTicker(24 * time.Hour)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ticker.C:
+				h.delete()
```

Comment:
why do we need to sleep another 24 hours before delete?

Reply:
yes,there is some problem,I change this to
```
defer ticker.Stop()
select {
	case <-ticker.C:
		ticker.Reset(24 * time.Hour)
		h.delete()
	case <-h.hotRegionInfoCtx.Done():
		return
}
for {
	select {
	   case <-ticker.C:
		h.delete()
	   case <-h.hotRegionInfoCtx.Done():
		return
	}
}
```